src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -504,12 +521,40 @@ public void logCurlCommand(String curlCommand) {
                     mCronetEngine,
                     mPool,
                     mUrlRewriter,
+                    mRequestListener,
                     mCurlLoggingEnabled,
                     mCurlCommandLogger,
                     mLogAuthTokensInCurlCommands);
         }
     }
 
+    /** Callback interface allowing clients to intercept different parts of the request flow. */
+    public abstract static class RequestListener {
+        private CronetHttpStack mStack;
+
+        void initialize(CronetHttpStack stack) {
+            mStack = stack;
+        }
+
+        /**
+         * Called when a request is prepared and about to be sent over the network.
+         *
+         * <p>Clients may use this callback to customize UrlRequests before they are dispatched,
+         * e.g. to enable socket tagging or request finished listeners.
+         */
+        public void onRequestPrepared(Request<?> request, UrlRequest.Builder requestBuilder) {}
+
+        /** @see AsyncHttpStack#getNonBlockingExecutor() */
+        protected Executor getNonBlockingExecutor() {
+            return mStack.getNonBlockingExecutor();
+        }
+
+        /** @see AsyncHttpStack#getBlockingExecutor() */
+        protected Executor getBlockingExecutor() {
+            return mStack.getBlockingExecutor();
+        }
```

Comment:
Why do we need this? Is it so `#onRequestPrepared` can choose to defer its execution to one of these executor? If so, shouldn't that be up to the application to decide which executor to use?

Reply:
I expect one common use case here to be calling ExperimentalUriRequest.Builder.setRequestFinishedListener, and subclassing RequestFinishedInfo.Listener requires providing the executor on which callbacks will be invoked later on.

I did think about keeping the interface simpler and just requiring clients to provide their own executors. While this is probably reasonable for clients that plan to provide their own (likely shared) executors as an input to Volley, keep in mind that Volley does provide default constructors with default thread pools to use. So if you're just using the default config, as I expect many applications will do, you have no way of accessing those executors and would have to create / provide a separate one just for this interface. Thus, we expose them here, just like we provide them in getters in AsyncHttpStack.

Reply:
Sounds good, thanks for the extra context